ChronoStore/src/storywriter.cpp

```diff
@@ -0,0 +1,497 @@
+storywriter::~storywriter(){}
+
+/**
+ * Write/Append data to storyDataset. 
+ * @param storyChunk: Data to write to story dataset.
+ * @param STORY: story name
+ * @param CHRONICLE: chronicle name
+ * @return: 0 if successful, else -1, if failed.
+ */
+int storywriter::writeStoryChunk(std::vector<Event> storyChunk, const char* STORY, const char* CHRONICLE) {
+
+    // Disable automatic printing of HDF5 error stack to console
+    if(!DEBUG){
+        H5Eset_auto(H5E_DEFAULT, NULL, NULL);
+    }
+
+    hid_t   ChronicleH5, storyDataset, storyDatasetSpace, chunkPropId, hyperslabMemorySpace, eventType, chunkmetaAttribute, attr, attr1, attr_DatasetMin, attr_DatasetMax, attr_DatasetTotalStoryEvents;
+    hid_t   attributeId, attributeType, attributeDataSpace;
+    herr_t  status, ret;
+
+    Event* eventBuffer;
+    ChunkAttr* attributeDataBuffer;
+    ChunkAttr* extendChunkMetadataBuffer;
+
+    /* 
+    * Chunk Dimension and number of events in a story chunk
+    */
+    const hsize_t chunkDims[1] = {CHUNK_SIZE};
+    const hsize_t numberOfEvents = storyChunk.size() - 1;
+
+    /*
+    * Allocate memory for events buffer
+    * TODO: Modification required in case of variable sized events
+    */
+    eventBuffer = (Event*) malloc(numberOfEvents * sizeof(Event));
+    if(eventBuffer == NULL){
+        return -1;
+    }
+
+    /*
+    * Allocate memory for attribute buffer
+    */
+    attributeDataBuffer = (ChunkAttr*) malloc(((numberOfEvents / CHUNK_SIZE) + 1) * sizeof(ChunkAttr));
+    if(attributeDataBuffer == NULL){
+        if(eventBuffer != NULL) free(eventBuffer);
+        return -1;
+    }
+
+    /*
+    * Read data from storyChunk into eventBuffer
+    */
+    for(auto event = 0 ; event < numberOfEvents ; event++) {
+    eventBuffer[event].timeStamp = storyChunk[event].timeStamp;
+    strcpy(eventBuffer[event].data, storyChunk[event].data);
+    }
+
+    /*
+    * Dimensions for the storyDataset
+    */
+    const hsize_t storyChunkDims[] = {storyChunk.size() - 1};
+
+    /*
+    * User defined compound datatype - consists "timeStamp" and "data"
+    */
+    herr_t timeStampError, dataError;
+    eventType = H5Tcreate(H5T_COMPOUND, sizeof(Event));
+    timeStampError = H5Tinsert(eventType, "timeStamp", HOFFSET(Event, timeStamp), H5T_NATIVE_ULONG);
+    dataError = H5Tinsert(eventType, "data", HOFFSET(Event, data), H5T_NATIVE_CHAR);
+    if(eventType < 0 || dataError < 0 || timeStampError < 0){
+        if(eventType >=0 )  H5Tclose(eventType);
+        if(attributeDataBuffer != NULL) free(attributeDataBuffer);
+        if(eventBuffer != NULL) free(eventBuffer);
+        return -1;
+    }
+
+    hsize_t storyDatasetSpaceStart = 0;
+    hsize_t newStorySize[1];
+
+    /*
+    * Check if the ChronicleH5 already exists
+    */
+    if (H5Fis_hdf5(CHRONICLE) > 0)
+    {
+        /*
+        * Open ChronicleH5 and check if the storyDataset exists
+        */
+
+        ChronicleH5 = H5Fopen(CHRONICLE, H5F_ACC_RDONLY, H5P_DEFAULT);
```

Comment:
Do we have to open it in read-only first, check existence and reopen in read-write mode if the dataset exists?

Reply:
Modified the code to open file only once in read-write. Thanks for pointing out.